File: Builder/python/Builder.py
from abc import ABCMeta


class Phone:

    def __init__(self):
        self.data: str = ''

    def about_phone(self) -> str:
        return self.data

    def append_data(self, string: str):
        self.data += string


class IDeveloper(metaclass=ABCMeta):

    def create_display(self):
        pass

    def create_box(self):
        pass

    def system_install(self):
        pass

    def get_phone(self) -> Phone:
        pass


class AndroidDeveloper(IDeveloper):

    def __init__(self):
        self.__phone = Phone()

    def create_display(self):
        self.__phone.append_data("(create) display Samsung; ")

    def create_box(self):
        self.__phone.append_data("(create) corpus Samsung; ")

    def system_install(self):
        self.__phone.append_data("(install) system Android; ")

    def get_phone(self) -> Phone:
        return self.__phone


class IphoneDeveloper(IDeveloper):

    def __init__(self):
        self.__phone = Phone()

    def create_display(self):
        self.__phone.append_data("(create) display Iphone; ")

    def create_box(self):
        self.__phone.append_data("(create) corpus Iphone; ")

    def system_install(self):
        self.__phone.append_data("(install) system IOS; ")

    def get_phone(self) -> Phone:
        return self.__phone


class Director:

    def __init__(self, developer: IDeveloper):
        self.__developer = developer

    def set_developer(self, developer: IDeveloper):
        self.__developer = developer

    def mount_only_phone(self) -> Phone:
        self.__developer.create_box()
        self.__developer.create_display()
        return self.__developer.get_phone()

    def mount_full_phone(self) -> Phone:
        self.__developer.create_box()
        self.__developer.create_display()
        self.__developer.system_install()
        return self.__developer.get_phone()


if __name__ == '__main__':
    
    android_developer: IDeveloper = AndroidDeveloper()

    director = Director(android_developer)

    samsung: Phone = director.mount_full_phone()
    print(samsung.about_phone())

    iphone_developer: IDeveloper = IphoneDeveloper()
    director.set_developer(iphone_developer)

    iphone: Phone = director.mount_only_phone()
    print(iphone.about_phone())

File: Builder/c++/Builder.cpp
#include <iostream>
#include <string>

class Phone
{
        std::string data;

    public:
        Phone() : data("") {}
        std::string aboutPhone();
        void appendData(std::string str);
};

class IDeveloper
{
    public:
        virtual void createDisplay() = 0;
        virtual void createBox() = 0;
        virtual void systemInstall() = 0;
        virtual Phone* GetPhone() = 0;
};

class AndroidDeveloper : public IDeveloper
{
        Phone* phone;

    public:
        AndroidDeveloper();
        ~AndroidDeveloper();
        void createDisplay() override;
        void createBox() override;
        void systemInstall() override;
        Phone* GetPhone() override;
};

class IPhoneDeveloper : public IDeveloper
{
        Phone* phone;

    public:
        IPhoneDeveloper();
        ~IPhoneDeveloper();
        void createDisplay() override;
        void createBox() override;
        void systemInstall() override;
        Phone* GetPhone() override;
};

class Director
{
        IDeveloper* developer;

    public:
        Director(IDeveloper* developer) : developer(developer) {}
        void SetDeveloper(IDeveloper* developer);
        Phone* MountOnlyPhone();
        Phone* MountFullPhone();
};

int main()
{
    IDeveloper* androidDeveloper = new AndroidDeveloper();
    Director director(androidDeveloper);

    Phone* samsung = director.MountFullPhone();
    std::cout << samsung->aboutPhone() << std::endl;

    IDeveloper* iphoneDeveloper = new IPhoneDeveloper();
    director.SetDeveloper(iphoneDeveloper);

    Phone* iphone = director.MountOnlyPhone();
    std::cout << iphone->aboutPhone() << std::endl;
    
    return 0;
}

std::string Phone::aboutPhone()
{
    return data;
}

void Phone::appendData(std::string str)
{
    data += str;
}

AndroidDeveloper::AndroidDeveloper()
{
    phone = new Phone();
}

AndroidDeveloper::~AndroidDeveloper()
{
    delete phone;
}

void AndroidDeveloper::createDisplay()
{
    phone->appendData("(create) display Samsung; ");
}

void AndroidDeveloper::createBox()
{
    phone->appendData("(create) corpus Samsung; ");
}

void AndroidDeveloper::systemInstall()
{
    phone->appendData("(install) system Android; ");
}

Phone *AndroidDeveloper::GetPhone()
{
    return phone;
}

IPhoneDeveloper::IPhoneDeveloper()
{
    phone = new Phone();
}

IPhoneDeveloper::~IPhoneDeveloper()
{
    delete phone;
}

void IPhoneDeveloper::createDisplay()
{
    phone->appendData("(create) display Iphone; ");
}

void IPhoneDeveloper::createBox()
{
    phone->appendData("(create) corpus Iphone; ");
}

void IPhoneDeveloper::systemInstall()
{
    phone->appendData("(install) system IOS; ");
}

Phone *IPhoneDeveloper::GetPhone()
{
    return phone;
}

void Director::SetDeveloper(IDeveloper *developer)
{
    this->developer = developer;
}

Phone *Director::MountOnlyPhone()
{
    developer->createBox();
    developer->createDisplay();
    return developer->GetPhone();
}

Phone *Director::MountFullPhone()
{
    developer->createBox();
    developer->createDisplay();
    developer->systemInstall();
    return developer->GetPhone();
}